pkg/querier/queryrange/results_cache.go

```diff
@@ -169,23 +201,32 @@ func (s resultsCache) Do(ctx context.Context, r Request) (Response, error) {
 
 // shouldCacheResponse says whether the response should be cached or not.
 func (s resultsCache) shouldCacheResponse(r Response) bool {
+	headerValues := getHeaderValuesWithName(r, cachecontrolHeader)
+	for _, v := range headerValues {
+		if v == noCacheValue {
+			level.Debug(s.logger).Log("msg", fmt.Sprintf("%s header in response is equal to %s, not caching the response", cachecontrolHeader, noCacheValue))
+			return false
+		}
+	}
+
+	return true
+}
+
+func getHeaderValuesWithName(r Response, headerName string) (headerValues []string) {
 	if promResp, ok := r.(*PrometheusResponse); ok {
 		for _, hv := range promResp.Headers {
 			if hv.GetName() != cachecontrolHeader {
 				continue
 			}
-			for _, v := range hv.GetValues() {
-				if v == noCacheValue {
-					level.Debug(s.logger).Log("msg", fmt.Sprintf("%s header in response is equal to %s, not caching the response", cachecontrolHeader, noCacheValue))
-					return false
-				}
-			}
+
+			headerValues = append(headerValues, hv.GetValues()...)
 		}
 	}
-	return true
+
+	return
 }
 
-func (s resultsCache) handleMiss(ctx context.Context, r Request) (Response, []Extent, error) {
+func (s resultsCache) handleMiss(ctx context.Context, r Request, userID string) (Response, []Extent, error) {
```

Comment:
```suggestion
func (s resultsCache) handleMiss(ctx context.Context, r Request) (Response, []Extent, error) {
```

The userID seems to be unused.